### ISSUE TITLE
All Games - Crash when opening Camera (Only on Windows, only with MinGW compiler)

### ISSUE THREAD

**Issue by acnleditor2** [citra-emu/citra#3051]
Commit: citra-emu/citra@79852d3707843a31f56b8f2dbc50fa91b4851be8
Stopped working on Nightly 373.
With only hw renderer, with hw renderer+shader jit, only shader jit, hw renderer+shader jit disabled.
Specs i5-7200U, Intel HD Graphics 620, Windows 10 15063.
Log (nightly 372, works):
https://pastebin.com/3mRE92d4

Log (new, dont works):
https://pastebin.com/4fWSUVtR

**Comment by wwylele** [comment]
I cannot reproduce. Same commit 26065fd, Ubuntu system, tested games:
 - Pokemon Sun (QR code)
 - Pokemon Super Mystery Dungeon (Intro without a save)
 - Mii Maker (QR code)
Black image as usual, without crashing

**Comment by adityaruplaha** [comment]
Reproduced by me.

https://community.citra-emu.org/t/psmd-crash-on-strtup-i-worked-well-earlier/4541

**Comment by adityaruplaha** [comment]
Happens on #b36f748 as well.

**Comment by wwylele** [comment]
Alright, here is a summery of some researching:

 - The crash comes from the `std::future`/`std::async` usage in the camera module. This is confirmed by commenting out all the `std::future` usage to see no crash anymore.
 - The stack is broken upon the crash
 - The crash only happens in the Qt frontend. Built in the same environment on the same commit, and running the same game, the SDL frontend doesn't crash
 - To test if `std::future` itself is the problem, an experiment was conducted:
the example code from http://en.cppreference.com/w/cpp/thread/future was put into
   - a standalone source file
   - the beginning of the `main()` in citra Qt frontend
   - the beginning of the `main()` in citra SDL frontend

  and compile them under the same msys environment. The result is that only the Qt frontend one crashes on boot, and the other two give the correct result
 - The theory is that `std::future` gets broken somehow by linking against Qt library
 - Also, the telemetry crash (#3039) has a very similar broken stack. Telemetry is also the other one of the only two user of `std::future` in citra. This indicates that it might be the crash comes from the same source.

**Comment by acnleditor2** [comment]
![sdl](https://user-images.githubusercontent.com/22228082/32138905-362b24ce-bc01-11e7-9ee8-24ba83dd7d66.PNG)
on sdl

**Comment by mailwl** [comment]
maybe use boost future in this case?

**Comment by wwylele** [comment]
So I have got some inconsistent report from others:
 - @Hexagon12 reports that inserting `std::future` test the Qt frontend doesn't crash and gives correct output
 - @acnleditor2 reports that nightly SDL frontend crashes on camera

I am not sure what this actually means, but probably Qt library is unrelated.

**Comment by wwylele** [comment]
A summery of my research today:

I managed to create a minimal crashing program:

```
#include <mutex>
#include <QtWidgets>
std::once_flag a;
int main(int argc, char* argv[]) {
    QApplication app(argc, argv);
    // std::call_once is used in the gcc implementation of `std::future`,
    // which causes the crash
    std::call_once(a, []() {});
}
```

Create a new subdirectory in citra with single source file of this. Link it against Qt and **any other one library**, for example
```
add_executable(crash_test ${SRCS})
target_link_libraries(crash_test PRIVATE Qt5::Widgets glad) # glad can be replaced by inih, or even an absolutely empty library
```
compile it and you will observe a crash.

Any **one of** the following modification will eliminate the crash:
 - remove `QApplication app(argc, argv);`
 - remove the linkage against other libraries (`glad` in the example above) and only leave Qt there

I also checked the output binary difference of the second modification. In the implementation of `std::call_once`, there is a call to `__emutls_get_address`. In the binary without linking additional libraries (which doesn't crash), this function directly calls to external dynamic library; however, in the binary with additional libraries linked (which crashes), this function is implemented in the binary (as if static linked). 

The function  `__emutls_get_address` apparently have something to do with TLS (thread local storage). My theory is that Qt also uses this function, and when the two "TLS manager"(?) instances get separated, by unknown bug when linking an additional libraries, calling to both will cause the crash (thus removing Qt invocation also eliminate the crash)

**Comment by adityaruplaha** [comment]
@wwylele I don't know much about coding but it looks like linking 2 libraries at the same time is causing the failure, right?

**Comment by adityaruplaha** [comment]
Believe it or not, it is working absolutely properly now, with a minor speed boost. I have commit `1fda9f`.

**Comment by acnleditor2** [comment]
not fixed...

**Comment by adityaruplaha** [comment]
Then why its working on my pc ? I played another chapter today...

**Comment by wwylele** [comment]
@adityaruplaha the crash only happens when you start without a save. Did you test it?

**Comment by adityaruplaha** [comment]
Ok ! That's the problem ! I'll test it.

**Comment by wwylele** [comment]
https://www.youtube.com/watch?v=xVT1y0xWgww
This might be relevant. both Qt, as a dynamic library, and static libraries in our project refers to the global emutls state, which might cause ctor/dtor/whatever duplication as shown in the video.

**Comment by wwylele** [comment]
New progress:

As previous summary, this change can eliminate the crash:
> remove the linkage against other libraries (glad in the example above) and only leave Qt there

The linker command generated by cmake changes upon this. More specifically, when another libraries gets link, the link command additionally links against `libgcc_eh`, while it doesn't link `libgcc_eh` if `glad`/`inih`/whatever gets linked. emutls stuff is defined in and exported by `libgcc_eh`.

**Comment by B3n30** [comment]
can we manually force to link against libgcc_eh?

**Comment by wwylele** [comment]
@B3n30 Sorry that I didn't make it clear... When libgcc_eh *is* linked, there *is* a crash. 
You can force link/unlink by modifying the makefile generated by cmake, but this is just a hack :p

**Comment by wwylele** [comment]
Root problem found: this is a cmake bug/regression (https://gitlab.kitware.com/cmake/cmake/issues/17436). This is fixed recently in cmake 3.9.6, and we just need to wait for AppVeyor's MSYS update (or can we update it manually?)
